### PR TITLE
Feature/rollback

### PR DESCRIPTION
**Pull request by Svastits**
Rollback is simplified for configuration and activation of the system_manager
- if one component fails, the cleanup/deactivation callback is called for the system manager
- these are implemented so, that they check the actual state of the components and do the rollback if necessary
- this way the nodes should be always in sync (if the rollback fails, error state is triggered)

Changed sync client requests to new SendRequest<> wrapper

### CODE REVIEW

File: kuka_sunrise_driver/src/kuka_sunrise/configuration_manager.cpp [hunk 1]
```diff
@@ -404,23 +404,17 @@ bool ConfigurationManager::setCommandMode(const std::string & control_mode)
     RCLCPP_ERROR(robot_manager_node_->get_logger(), "Invalid control mode");
     return false;
   }
-  auto future_result = command_mode_client_->async_send_request(request);
```
Comment (kovacsge11) [inline kuka_sunrise_driver/src/kuka_sunrise/configuration_manager.cpp:407]: I don't know where to put it, but we should refactor parameter handling in this source file, too. Right?

File: kuka_sunrise_driver/src/kuka_sunrise/robot_manager_node.cpp [hunk 2]
```diff
@@ -97,19 +97,14 @@ RobotManagerNode::on_configure(const rclcpp_lifecycle::State &)
...
-    std::chrono::milliseconds(3000));
-  if (future_status != std::future_status::ready) {
-    RCLCPP_ERROR(get_logger(), "Future status not ready, could not set parameters");
-    return FAILURE;
-  }
-  if (!future_result.get()->success) {
-    RCLCPP_ERROR(get_logger(), "Future result not success, could not set parameters");
+
+  auto response = kuka_sunrise::sendRequest<std_srvs::srv::Trigger::Response>(
+    set_parameter_client_, trigger_request, 0, 1000);
```
Comment (kovacsge11) [inline kuka_sunrise_driver/src/kuka_sunrise/robot_manager_node.cpp:102]: Why did you change timeout ms from 3000 to 1000? It occurs multiple times in this PR. (I mean it's ok, but I don't know whether you changed it on purpose.)

File: kuka_sunrise_control/teleop_guided_robot/package.xml [hunk 3]
```diff
@@ -3,9 +3,9 @@
 <package format="3">
   <name>teleop_guided_robot</name>
   <version>0.0.0</version>
-  <description>TODO: Package description</description>
+  <description>Package for managing the connection between ROS and KUKA Sunrise FRI</description>
```
Comment (kovacsge11) [inline kuka_sunrise_control/teleop_guided_robot/package.xml:6]: But this is not the purpose of the teleop_guided_robot package, is it?

**Review by kovacsge11 — CHANGES_REQUESTED** [review]
Review not finished yet, system_manager class needs review.

File: kuka_sunrise_driver/src/kuka_sunrise/robot_manager_node.cpp (hunk 2 above)
Comment (Svastits) [reply kuka_sunrise_driver/src/kuka_sunrise/robot_manager_node.cpp:102]: For the changes of simple values (member variables or FRI parameters), 1000ms is more than enough, so I changed them on purpose.
Only the requests of state changes need a longer time, so I left those on the original 3000 ms

File: kuka_sunrise_control/teleop_guided_robot/package.xml (hunk 3 above)
Comment (Svastits) [reply kuka_sunrise_control/teleop_guided_robot/package.xml:6]: True, I will update this
But I think we should consider moving the system_manager to another package, as it has a more general purpose than just starting the teleop_guided motion

File: kuka_sunrise_driver/src/kuka_sunrise/configuration_manager.cpp (hunk 1 above)
Comment (Svastits) [reply kuka_sunrise_driver/src/kuka_sunrise/configuration_manager.cpp:407]: Yes, that should be done, but I would do it in a separate PR, as the current ParameterSetAccessRights structure of the base node is not sufficient for this purpose:
- in this case, the ConfigurationManager is instantiated while the node is in configuring (transition) state, but currently only primary states are supported in that struct
- either the struct should be modified to enable transition states or the code restructured -> FSM restructuring should be planned before

File: kuka_sunrise_control/teleop_guided_robot/src/control_system/system_manager.cpp [hunk 4]
```diff
@@ -14,107 +14,211 @@
...
+  {
     return ERROR;
   }
-  if (!changeState(JOINT_CONTROLLER, lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE)) {
+  if (getState(JOINT_CONTROLLER).id != lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE &&
+    !changeState(
+      JOINT_CONTROLLER,
+      lifecycle_msgs::msg::Transition::TRANSITION_DEACTIVATE))
+  {
+    return ERROR;
```
Comment (kovacsge11) [inline kuka_sunrise_control/teleop_guided_robot/src/control_system/system_manager.cpp:204]: why use ERROR here and FAILURE at other places?

File: kuka_sunrise_control/teleop_guided_robot/src/control_system/system_manager.cpp [hunk 5]
```diff
@@ -14,107 +14,211 @@
...
 {
-  (void)state;
-  if (!changeState(ROBOT_INTERFACE, lifecycle_msgs::msg::Transition::TRANSITION_CLEANUP)) {
-    return FAILURE;
+  if (getState(ROBOT_INTERFACE).id != lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED &&
+    !changeState(
+      ROBOT_INTERFACE,
+      lifecycle_msgs::msg::Transition::TRANSITION_CLEANUP))
+  {
+    return ERROR;
```
Comment (kovacsge11) [inline kuka_sunrise_control/teleop_guided_robot/src/control_system/system_manager.cpp:126]: same here, why ERROR instead of FAILURE?

**Review by kovacsge11 — APPROVED** [review]

**Review by kovacsge11 — CHANGES_REQUESTED** [review]
ERROR/FAILURE not clear

File: kuka_sunrise_control/teleop_guided_robot/src/control_system/system_manager.cpp (hunk 5 above)
Comment (Svastits) [reply kuka_sunrise_control/teleop_guided_robot/src/control_system/system_manager.cpp:126]: For deactivation and cleanup, I felt it is safer to return error if transition is not successful, in these cases, the node will go to error state and get finalized or return to unconfigured.

For configure or activate, this is not the desired behaviour, it is safe to return to the original state and thus enable retrying the transition

**Review by kovacsge11 — APPROVED** [review]
